File: lib/features/booking/data/models/booking.dart
import '../../../../core/constants/app_constants.dart';

class ContactInfo {
  final String email;
  final String phone;
  final String? firstName;
  final String? lastName;

  const ContactInfo({
    required this.email,
    required this.phone,
    this.firstName,
    this.lastName,
  });

  factory ContactInfo.fromJson(Map<String, dynamic> json) {
    return ContactInfo(
      email: json['email'] ?? '',
      phone: json['phone'] ?? '',
      firstName: json['firstName'],
      lastName: json['lastName'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'email': email,
      'phone': phone,
      'firstName': firstName,
      'lastName': lastName,
    };
  }

  ContactInfo copyWith({
    String? email,
    String? phone,
    String? firstName,
    String? lastName,
  }) {
    return ContactInfo(
      email: email ?? this.email,
      phone: phone ?? this.phone,
      firstName: firstName ?? this.firstName,
      lastName: lastName ?? this.lastName,
    );
  }
}

class Booking {
  final String id;
  final String userId;
  final String scheduleId;
  final String pnr;
  final List<Passenger> passengers;
  final String selectedClass;
  final List<String> selectedSeats;
  final BookingStatus status;
  final double totalPrice;
  final String currency;
  final ContactInfo contactInfo;
  final String paymentMethod;
  final String? paymentId;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final DateTime? confirmedAt;
  final DateTime? cancelledAt;
  final String? cancelReason;
  final DateTime? expiresAt;
  final Map<String, dynamic> metadata;

  const Booking({
    required this.id,
    required this.userId,
    required this.scheduleId,
    required this.pnr,
    required this.passengers,
    required this.selectedClass,
    required this.selectedSeats,
    required this.status,
    required this.totalPrice,
    required this.currency,
    required this.contactInfo,
    required this.paymentMethod,
    this.paymentId,
    required this.createdAt,
    this.updatedAt,
    this.confirmedAt,
    this.cancelledAt,
    this.cancelReason,
    this.expiresAt,
    this.metadata = const {},
  });

  /// Helper method to extract ID from object or string
  static String _extractId(dynamic value) {
    if (value == null) return '';
    if (value is String) return value;
    if (value is Map<String, dynamic>) {
      return value['_id'] ?? value['id'] ?? '';
    }
    return '';
  }

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['_id'] ?? json['id'] ?? '',
      userId: _extractId(json['userId']),
      scheduleId: _extractId(json['scheduleId']),
      pnr: json['pnr'] ?? '',
      passengers:
          (json['passengers'] as List?)
              ?.map((e) => Passenger.fromJson(e))
              .toList() ??
          [],
      selectedClass: json['selectedClass'] ?? '',
      selectedSeats: List<String>.from(json['selectedSeats'] ?? []),
      status: BookingStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => BookingStatus.pending,
      ),
      totalPrice: (json['totalPrice'] ?? 0).toDouble(),
      currency: json['currency'] ?? 'VND',
      contactInfo: json['contactInfo'] != null
          ? ContactInfo.fromJson(json['contactInfo'])
          : const ContactInfo(email: '', phone: ''),
      paymentMethod: json['paymentMethod'] ?? '',
      paymentId: json['paymentId'],
      createdAt: DateTime.parse(
        json['createdAt'] ?? DateTime.now().toIso8601String(),
      ),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
      confirmedAt: json['confirmedAt'] != null
          ? DateTime.parse(json['confirmedAt'])
          : null,
      cancelledAt: json['cancelledAt'] != null
          ? DateTime.parse(json['cancelledAt'])
          : null,
      cancelReason: json['cancelReason'],
      expiresAt: json['expiresAt'] != null
          ? DateTime.parse(json['expiresAt'])
          : null,
      metadata: Map<String, dynamic>.from(json['metadata'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'scheduleId': scheduleId,
      'pnr': pnr,
      'passengers': passengers.map((e) => e.toJson()).toList(),
      'selectedClass': selectedClass,
      'selectedSeats': selectedSeats,
      'status': status.name,
      'totalPrice': totalPrice,
      'currency': currency,
      'contactInfo': contactInfo.toJson(),
      'paymentMethod': paymentMethod,
      'paymentId': paymentId,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'confirmedAt': confirmedAt?.toIso8601String(),
      'cancelledAt': cancelledAt?.toIso8601String(),
      'cancelReason': cancelReason,
      'expiresAt': expiresAt?.toIso8601String(),
      'metadata': metadata,
    };
  }

  Booking copyWith({
    String? id,
    String? userId,
    String? scheduleId,
    String? pnr,
    List<Passenger>? passengers,
    String? selectedClass,
    List<String>? selectedSeats,
    BookingStatus? status,
    double? totalPrice,
    String? currency,
    ContactInfo? contactInfo,
    String? paymentMethod,
    String? paymentId,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? confirmedAt,
    DateTime? cancelledAt,
    String? cancelReason,
    DateTime? expiresAt,
    Map<String, dynamic>? metadata,
  }) {
    return Booking(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      scheduleId: scheduleId ?? this.scheduleId,
      pnr: pnr ?? this.pnr,
      passengers: passengers ?? this.passengers,
      selectedClass: selectedClass ?? this.selectedClass,
      selectedSeats: selectedSeats ?? this.selectedSeats,
      status: status ?? this.status,
      totalPrice: totalPrice ?? this.totalPrice,
      currency: currency ?? this.currency,
      contactInfo: contactInfo ?? this.contactInfo,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      paymentId: paymentId ?? this.paymentId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      confirmedAt: confirmedAt ?? this.confirmedAt,
      cancelledAt: cancelledAt ?? this.cancelledAt,
      cancelReason: cancelReason ?? this.cancelReason,
      expiresAt: expiresAt ?? this.expiresAt,
      metadata: metadata ?? this.metadata,
    );
  }
}

class Passenger {
  final String id;
  final String firstName;
  final String lastName;
  final PassengerType type;
  final String documentType;
  final String documentNumber;
  final DateTime? dateOfBirth;
  final String? nationality;
  final String? gender;
  final String? seatNumber;
  final Map<String, dynamic> metadata;

  const Passenger({
    required this.id,
    required this.firstName,
    required this.lastName,
    required this.type,
    required this.documentType,
    required this.documentNumber,
    this.dateOfBirth,
    this.nationality,
    this.gender,
    this.seatNumber,
    this.metadata = const {},
  });

  factory Passenger.fromJson(Map<String, dynamic> json) {
    return Passenger(
      id: json['_id'] ?? json['id'] ?? '',
      firstName: json['firstName'] ?? '',
      lastName: json['lastName'] ?? '',
      type: PassengerType.values.firstWhere(
        (e) => e.name == json['type'],
        orElse: () => PassengerType.adult,
      ),
      documentType: json['documentType'] ?? '',
      documentNumber: json['documentNumber'] ?? '',
      dateOfBirth: json['dateOfBirth'] != null
          ? DateTime.parse(json['dateOfBirth'])
          : null,
      nationality: json['nationality'],
      gender: json['gender'],
      seatNumber: json['seatNumber'],
      metadata: Map<String, dynamic>.from(json['metadata'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'firstName': firstName,
      'lastName': lastName,
      'type': type.name,
      'documentType': documentType,
      'documentNumber': documentNumber,
      'dateOfBirth': dateOfBirth?.toIso8601String(),
      'nationality': nationality,
      'gender': gender,
      'seatNumber': seatNumber,
      'metadata': metadata,
    };
  }
}

class BookingRequest {
  final String scheduleId;
  final List<Passenger> passengers;
  final String selectedClass;
  final List<String> selectedSeats;
  final ContactInfo contactInfo;
  final String paymentMethod;
  final Map<String, dynamic> metadata;

  const BookingRequest({
    required this.scheduleId,
    required this.passengers,
    required this.selectedClass,
    required this.selectedSeats,
    required this.contactInfo,
    required this.paymentMethod,
    this.metadata = const {},
  });

  factory BookingRequest.fromJson(Map<String, dynamic> json) {
    return BookingRequest(
      scheduleId: json['scheduleId'] ?? '',
      passengers:
          (json['passengers'] as List?)
              ?.map((e) => Passenger.fromJson(e))
              .toList() ??
          [],
      selectedClass: json['selectedClass'] ?? '',
      selectedSeats: List<String>.from(json['selectedSeats'] ?? []),
      contactInfo: ContactInfo.fromJson(json['contactInfo'] ?? {}),
      paymentMethod: json['paymentMethod'] ?? 'vnpay',
      metadata: Map<String, dynamic>.from(json['metadata'] ?? {}),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'scheduleId': scheduleId,
      'passengers': passengers.map((e) => e.toJson()).toList(),
      'selectedClass': selectedClass,
      'selectedSeats': selectedSeats,
      'contactInfo': contactInfo.toJson(),
      'paymentMethod': paymentMethod,
      'metadata': metadata,
    };
  }
}

File: backend/src/models/Booking.js
const mongoose = require('mongoose');

const PassengerSchema = new mongoose.Schema({
  type: {
    type: String,
    enum: ['adult', 'child', 'infant'],
    required: true
  },
  firstName: {
    type: String,
    required: true,
    trim: true
  },
  lastName: {
    type: String,
    required: true,
    trim: true
  },
  dateOfBirth: Date,
  gender: {
    type: String,
    enum: ['male', 'female']
  },
  documentType: {
    type: String,
    enum: ['passport', 'id_card', 'driver_license'],
    required: true
  },
  documentNumber: {
    type: String,
    required: true,
    trim: true
  },
  nationality: String,
  seatNumber: String,
  metadata: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  }
}, { _id: true });

const ContactInfoSchema = new mongoose.Schema({
  email: {
    type: String,
    required: true,
    lowercase: true,
    trim: true
  },
  phone: {
    type: String,
    required: true,
    trim: true
  },
  firstName: String,
  lastName: String
}, { _id: false });

const BookingSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  scheduleId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Schedule',
    required: true,
    index: true
  },
  pnr: {
    type: String,
    required: true,
    unique: true,
    uppercase: true,
    index: true
  },
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'cancelled', 'completed', 'expired'],
    default: 'pending',
    index: true
  },
  passengers: [PassengerSchema],
  selectedClass: {
    type: String,
    required: true
  },
  selectedSeats: [String],
  totalPrice: {
    type: Number,
    required: true,
    min: 0
  },
  currency: {
    type: String,
    default: 'VND',
    enum: ['VND', 'USD']
  },
  contactInfo: ContactInfoSchema,
  paymentMethod: {
    type: String,
    enum: ['vnpay', 'momo', 'stripe', 'bank_transfer'],
    required: true
  },
  paymentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Payment'
  },
  confirmedAt: Date,
  cancelledAt: Date,
  cancelReason: String,
  expiresAt: {
    type: Date,
    default: function() {
      return new Date(Date.now() + 30 * 60 * 1000); // 30 minutes
    },
    index: { expireAfterSeconds: 0 }
  },
  metadata: {
    type: mongoose.Schema.Types.Mixed,
    default: {}
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Indexes (pnr and expiresAt already have indexes defined in schema)
BookingSchema.index({ userId: 1, createdAt: -1 });
// BookingSchema.index({ pnr: 1 }); // Removed - already unique
BookingSchema.index({ status: 1, createdAt: -1 });
BookingSchema.index({ scheduleId: 1, status: 1 });
// BookingSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 }); // Removed - already defined in schema

// Virtuals
BookingSchema.virtual('isPending').get(function() {
  return this.status === 'pending';
});

BookingSchema.virtual('isConfirmed').get(function() {
  return this.status === 'confirmed';
});

BookingSchema.virtual('isCancelled').get(function() {
  return this.status === 'cancelled';
});

BookingSchema.virtual('passengerCount').get(function() {
  return this.passengers.length;
});

BookingSchema.virtual('adultCount').get(function() {
  return this.passengers.filter(p => p.type === 'adult').length;
});

BookingSchema.virtual('childCount').get(function() {
  return this.passengers.filter(p => p.type === 'child').length;
});

BookingSchema.virtual('infantCount').get(function() {
  return this.passengers.filter(p => p.type === 'infant').length;
});

BookingSchema.virtual('isExpired').get(function() {
  return this.status === 'pending' && new Date() > this.expiresAt;
});

// Methods
BookingSchema.methods.confirm = function() {
  this.status = 'confirmed';
  this.confirmedAt = new Date();
  this.expiresAt = undefined;
  return this.save();
};

BookingSchema.methods.cancel = function(reason) {
  this.status = 'cancelled';
  this.cancelledAt = new Date();
  this.cancelReason = reason;
  this.expiresAt = undefined;
  return this.save();
};

BookingSchema.methods.complete = function() {
  this.status = 'completed';
  return this.save();
};

BookingSchema.methods.extendExpiry = function(minutes = 30) {
  if (this.status === 'pending') {
    this.expiresAt = new Date(Date.now() + minutes * 60 * 1000);
    return this.save();
  }
  return Promise.resolve(this);
};

BookingSchema.methods.addPassenger = function(passenger) {
  this.passengers.push(passenger);
  return this.save();
};

BookingSchema.methods.removePassenger = function(passengerId) {
  this.passengers = this.passengers.filter(p => p._id.toString() !== passengerId.toString());
  return this.save();
};

BookingSchema.methods.updateSeat = function(passengerId, seatNumber) {
  const passenger = this.passengers.id(passengerId);
  if (passenger) {
    passenger.seatNumber = seatNumber;
    return this.save();
  }
  return Promise.reject(new Error('Passenger not found'));
};

BookingSchema.methods.generateTickets = async function() {
  const Ticket = require('./Ticket');
  const tickets = [];
  
  for (const passenger of this.passengers) {
    const ticketData = {
      bookingId: this._id,
      userId: this.userId,
      passengerId: passenger._id,
      pnr: this.pnr,
      ticketNumber: generateTicketNumber(),
      qrData: await generateQRData(this, passenger),
      status: 'issued'
    };
    
    const ticket = await Ticket.create(ticketData);
    tickets.push(ticket);
  }
  
  return tickets;
};

// Static methods
BookingSchema.statics.findByPNR = function(pnr) {
  return this.findOne({ pnr: pnr.toUpperCase() })
    .populate('userId', 'displayName email phoneNumber')
    .populate('scheduleId');
};

BookingSchema.statics.findByUser = function(userId, options = {}) {
  const query = this.find({ userId })
    .populate('scheduleId')
    .sort({ createdAt: -1 });
    
  if (options.status) {
    query.where({ status: options.status });
  }
  
  if (options.limit) {
    query.limit(options.limit);
  }
  
  return query;
};

BookingSchema.statics.findExpired = function() {
  return this.find({
    status: 'pending',
    expiresAt: { $lt: new Date() }
  });
};

BookingSchema.statics.findBySchedule = function(scheduleId) {
  return this.find({ 
    scheduleId, 
    status: { $in: ['confirmed', 'completed'] }
  }).populate('userId', 'displayName email');
};

// Pre-save middleware
BookingSchema.pre('save', function(next) {
  if (this.isNew && !this.pnr) {
    this.pnr = generatePNR();
  }
  next();
});

// Helper functions
function generatePNR() {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let pnr = '';
  for (let i = 0; i < 6; i++) {
    pnr += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return pnr;
}

function generateTicketNumber() {
  const timestamp = Date.now().toString().slice(-8);
  const random = Math.floor(Math.random() * 1000).toString().padStart(3, '0');
  return `DV${timestamp}${random}`;
}

async function generateQRData(booking, passenger) {
  const qrData = {
    bookingId: booking._id,
    pnr: booking.pnr,
    passengerName: `${passenger.firstName} ${passenger.lastName}`,
    passengerId: passenger._id,
    seatNumber: passenger.seatNumber,
    issuedAt: Date.now()
  };
  
  return JSON.stringify(qrData);
}

module.exports = mongoose.model('Booking', BookingSchema);
